### PR TITLE
Change VolumeStore to use digests instead of names on the filesystem

### PR DESCRIPTION
**Pull request by apostasie**
The motivation for this comes from #2942 - with some discussion in #3279

Restricting names to what the filesystem can do is very limiting.
Furthermore, user-controlled strings being used directly as paths components on the filesystem gives me the creeps.

This proposal here moves away from storing volumes inside `volume_root/volume_name` to `volume_root/digest(volume_name)`.

It is fully backward compatible: if a volume exists with the legacy layout, it will still be usable.

It is NOT entirely forward compatible though: although existing volumes will keep working even with older versions even after having been used by newer version, this is not true for entirely new volumes:
- for the older version, newly created volumes will have their name appear as a digest
- legacy volumes that have been touched by the new version will have a doppelganger which will appear with a name that is the digest

What do you folks think?

cc @AkihiroSuda and @Shubhranshu153 

FWIW docker IS using names as path components, and their regexp is: `[a-zA-Z0-9][a-zA-Z0-9_.-]`
They do enforce a limitation at max 254 characters as far as I can tell.

### CODE REVIEW

File: pkg/mountutil/volumestore/volumestore.go [hunk 1]
```diff
@@ -22,20 +22,38 @@ import (
...
 const (
 	dataDirName        = "_data"
 	volumeJSONFileName = "volume.json"
 )
 
+var (
+	validateRegExp = regexp.MustCompile("^[a-zA-Z0-9_+:/-]{1,256}$")
+)
+
+func validateName(name string) error {
```
Comment (apostasie) [inline pkg/mountutil/volumestore/volumestore.go:46]: Purely provisional ^ - to be replaced by whatever we deem right.
I do not quite care what this does exactly, as this PR technically should free us from any limitations on names.

File: pkg/mountutil/volumestore/volumestore.go [hunk 2]
```diff
@@ -105,19 +123,63 @@ func (vs *volumeStore) Unlock() error {
 	return nil
 }
 
+func (vs *volumeStore) legacyEnsureLink(name string) error {
```
Comment (apostasie) [inline pkg/mountutil/volumestore/volumestore.go:126]: If we have a legacy volume, calling this method will add the name inside `volFilePath` and create a symlink from `legacyVolPath` (with name) to `volPath` (with digest).

File: pkg/mountutil/volumestore/volumestore.go [hunk 3]
```diff
@@ -176,40 +241,62 @@ func (vs *volumeStore) Create(name string, labels []string) (*native.Volume, err
 
 // Get retrieves a native volume from the store
 // Besides a possible locking error, it might return ErrInvalidArgument, ErrNotFound, or a filesystem error
-func (vs *volumeStore) Get(name string, size bool) (*native.Volume, error) {
-	if err := identifiers.Validate(name); err != nil {
-		return nil, fmt.Errorf("malformed volume name %q: %w", name, err)
+func (vs *volumeStore) Get(nameOrDigest string, size bool) (*native.Volume, error) {
```
Comment (apostasie) [inline pkg/mountutil/volumestore/volumestore.go:244]: Given how `List` is working, `Get` will now be able to retrieve from either name or digest.

**Comment by Shubhranshu153** [comment]
From what i tested it doesnt seem docker is controlling it, rather the mkdir is failing for volume create
```
docker volume create --name ksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkk
Error response from daemon: create ksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkk: error while creating volume root path '/local/docker/volumes/ ksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkk': mkdir /local/docker/volumes/ksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkkksafnlksnaslknglksanglknglkkkk: file name too long
```

for other identifier names like containers i am able to give a pretty long name. (> 255 chars)


having a sha256 as identifier i feel needs to be a 2 way hash rather compression otherwise it would be difficult to list it out. Not sure if we need to complicate it.

**Comment by apostasie** [comment]
> for other identifier names like containers i am able to give a pretty long name. (> 255 chars)

Our NameStore implementation also uses the filesystem, so, same thing here: we need to rewrite it if we want to open it up.

> having a sha256 as identifier i feel needs to be a 2 way hash rather compression otherwise it would be difficult to list it out. Not sure if we need to complicate it.

`VolumeStore.Get` was already reading a data file (for labels) - just adding the name was trivial.
As for `List`, it does call `Get`, so.
The extra complexity here is having to deal with the legacy format transparently.

I do not see how we could lift our length limit with a two way hash (which I assume would have to be at least as long as the name it is hashing...).

**Comment by AkihiroSuda** [comment]
Thanks, but I feel this is overengineering and hard to test

**Comment by AkihiroSuda** [comment]
Something bad may happen when a legacy volume exists with hexadecimal name?

**Comment by apostasie** [comment]
> Something bad may happen when a legacy volume exists with hexadecimal name?

It would fail `digest.Parse()`, and be treated as "legacy" name.
Since we do right now limit at 76 chars, 

The only case where we could possibly run into trouble is if someone has been using as a volume name a valid digest prefixed by sha256 (eg: `nerdctl volume create sha256_b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c`).

If this is a concern, we could certainly adapt the new layout to be over 76 characters long. Eg:
```
nerdctl volume create my_volume_name
# ===> on the filesystem: volume_sha256_93ff9de5cd31a50432dd2e39012e53ae78dd9a7ccb520c04f93828873813fef8
```

Since legacy identifiers cannot be over 76 chars, that would do it.

**Comment by apostasie** [comment]
> Thanks, but I feel this is overengineering and hard to test

Ok, that's fine.

**Pull request closed without merging**